File: Editorial/0089_Gray_Code/JavaScript/4_Recursion_2.js
let nextNum = 0;
var grayCode = function (n) {
    let result = [];
    grayCodeHelper(result, n);
    nextNum = 0; // reset global variable
    return result;
};
let grayCodeHelper = function (result, n) {
    if (n == 0) {
        result.push(nextNum);
        return;
    }
    grayCodeHelper(result, n - 1);
    // Flip the bit at (n - 1)th position from right
    nextNum = nextNum ^ (1 << (n - 1));
    grayCodeHelper(result, n - 1);
};

File: Editorial/0089_Gray_Code/Go/4_Recursion_2.go
var nextNum int

func grayCode(n int) []int {
    var result []int
    nextNum = 0
    grayCodeHelper(&result, n)
    return result
}

func grayCodeHelper(result *[]int, n int) {
    if n == 0 {
        *result = append(*result, nextNum)
        return
    }
    grayCodeHelper(result, n-1)
    // Flip the bit at (n - 1)th position from right
    nextNum = nextNum ^ (1 << (n - 1))
    grayCodeHelper(result, n-1)
}
